### ISSUE TITLE
InvalidCastException 

### ISSUE THREAD

**Issue by bslbckr** [agileobjects/AgileMapper#133]
I observe an `InvalidCastException` while mapping a wrapper-object to a dictionary. It seems to be related to the issues #110 , however as I still get this exception using 1.4-preview2...

I have a MWE to reproduce the error:
```
namespace Source
{
    public class Wrapper
    {
         public IDictionary<string, Data> Map {get;} = new Dictionary<string, Data>();
     }
     public class Data
     {
        public string Name { get; set; }
     }
}

namespace Target
{
    public class Data
    {
        public string Name { get; set; }
    }
}
public class Program {
  public static int Main(string[] args) {
    Mapper.WhenMapping.From<Source.Wrapper>.To<Dictionary<string, Target.Data>>().Map(ctxt => ctxt.Source.Map).ToTarget();
    var input = new Source.Wrapper();
    input.Map.Add("first", new Source.Data{ Name = "First"});
    input.Map.Add("second", new Source.Data{ Name = "Second" });
    Mapper.Map(input).ToANew<Dictionary<string, Target.Data>>();
  }
}
```

If I run this snippet I receive the following stack-trace:
```
System.InvalidCastException : Unable to cast object of type 'AgileObjects.AgileMapper.Members.QualifiedMember' to type 'AgileObjects.AgileMapper.Members.Dictionaries.DictionaryTargetMember'.
[24.04.2019 11:13:28 Informational] [xUnit.net 00:00:01.09]       Stack Trace:
[24.04.2019 11:13:28 Informational] [xUnit.net 00:00:01.09]            at AgileObjects.AgileMapper.Members.Dictionaries.DictionaryTargetMember.Append(DictionaryMemberKey memberKey)
[24.04.2019 11:13:28 Informational] [xUnit.net 00:00:01.09]            at AgileObjects.AgileMapper.Members.Dictionaries.DictionaryTargetMember.Append(ParameterExpression key)
[24.04.2019 11:13:28 Informational] [xUnit.net 00:00:01.09]            at AgileObjects.AgileMapper.ObjectPopulation.Enumerables.Dictionaries.DictionaryPopulationBuilder.AssignDictionaryEntryFromKeyValuePair(DictionaryToDictionaryPopulationLoopData loopData, IObjectMappingData mappingData)
[24.04.2019 11:13:28 Informational] [xUnit.net 00:00:01.09]            at AgileObjects.AgileMapper.ObjectPopulation.Enumerables.PopulationLoopDataExtensions.BuildPopulationLoop[TLoopData](TLoopData loopData, EnumerablePopulationBuilder builder, IObjectMappingData mappingData, Func`3 elementPopulationFactory)
[24.04.2019 11:13:28 Informational] [xUnit.net 00:00:01.09]            at AgileObjects.AgileMapper.ObjectPopulation.Enumerables.Dictionaries.DictionaryPopulationBuilder.BuildDictionaryToDictionaryPopulationLoop(IObjectMappingData mappingData)
[24.04.2019 11:13:28 Informational] [xUnit.net 00:00:01.09]            at AgileObjects.AgileMapper.ObjectPopulation.DictionaryMappingExpressionFactory.GetEnumerableToDictionaryMapping(IObjectMappingData mappingData)
[24.04.2019 11:13:28 Informational] [xUnit.net 00:00:01.09]            at AgileObjects.AgileMapper.ObjectPopulation.DictionaryMappingExpressionFactory.GetObjectPopulation(MappingCreationContext context)+MoveNext()
[24.04.2019 11:13:28 Informational] [xUnit.net 00:00:01.09]            at AgileObjects.AgileMapper.Extensions.PublicEnumerableExtensions.Filter[TItem](IEnumerable`1 items, Func`2 predicate)+MoveNext()
[24.04.2019 11:13:28 Informational] [xUnit.net 00:00:01.09]            at System.Collections.Generic.List`1.AddEnumerable(IEnumerable`1 enumerable)
[24.04.2019 11:13:28 Informational] [xUnit.net 00:00:01.09]            at System.Collections.Generic.List`1.InsertRange(Int32 index, IEnumerable`1 collection)
[24.04.2019 11:13:28 Informational] [xUnit.net 00:00:01.09]            at AgileObjects.AgileMapper.ObjectPopulation.MappingExpressionFactoryBase.AddPopulationsAndCallbacks(MappingCreationContext context)
[24.04.2019 11:13:28 Informational] [xUnit.net 00:00:01.09]            at AgileObjects.AgileMapper.ObjectPopulation.MappingExpressionFactoryBase.Create(IObjectMappingData mappingData)
[24.04.2019 11:13:28 Informational] [xUnit.net 00:00:01.09]            at AgileObjects.AgileMapper.ObjectPopulation.ObjectMapperFactory.Create[TSource,TTarget](ObjectMappingData`2 mappingData)
[24.04.2019 11:13:28 Informational] [xUnit.net 00:00:01.09]            at AgileObjects.AgileMapper.ObjectPopulation.ObjectMappingData`2.GetOrCreateMapper()
[24.04.2019 11:13:28 Informational] [xUnit.net 00:00:01.09]            at AgileObjects.AgileMapper.ObjectPopulation.MappingFactory.GetInlineMappingBlock(IObjectMappingData mappingData, MappingValues mappingValues, Expression createMappingDataCall)
[24.04.2019 11:13:28 Informational] [xUnit.net 00:00:01.09]            at AgileObjects.AgileMapper.DataSources.Finders.DataSourceFindContext.GetFinalDataSource(IDataSource foundDataSource, IChildMemberMappingData mappingData)
[24.04.2019 11:13:28 Informational] [xUnit.net 00:00:01.09]            at AgileObjects.AgileMapper.DataSources.Finders.ConfiguredDataSourceFinder.FindFor(DataSourceFindContext context)+MoveNext()
[24.04.2019 11:13:28 Informational] [xUnit.net 00:00:01.09]            at AgileObjects.AgileMapper.DataSources.Finders.DataSourceFinder.EnumerateDataSources(DataSourceFindContext context)+MoveNext()
[24.04.2019 11:13:28 Informational] [xUnit.net 00:00:01.09]            at System.Collections.Generic.LargeArrayBuilder`1.AddRange(IEnumerable`1 items)
[24.04.2019 11:13:28 Informational] [xUnit.net 00:00:01.09]            at System.Collections.Generic.EnumerableHelpers.ToArray[T](IEnumerable`1 source)
[24.04.2019 11:13:28 Informational] [xUnit.net 00:00:01.09]            at System.Linq.Enumerable.ToArray[TSource](IEnumerable`1 source)
[24.04.2019 11:13:28 Informational] [xUnit.net 00:00:01.09]            at AgileObjects.AgileMapper.DataSources.Finders.DataSourceFinder.FindFor(IChildMemberMappingData childMappingData)
[24.04.2019 11:13:28 Informational] [xUnit.net 00:00:01.09]            at AgileObjects.AgileMapper.Members.Population.MemberPopulatorFactory.Create(QualifiedMember targetMember, IObjectMappingData mappingData)
[24.04.2019 11:13:28 Informational] [xUnit.net 00:00:01.09]            at AgileObjects.AgileMapper.Members.Population.MemberPopulatorFactory.<>c__DisplayClass3_0.<Create>b__0(QualifiedMember tm)
[24.04.2019 11:13:28 Informational] [xUnit.net 00:00:01.09]            at AgileObjects.AgileMapper.Extensions.PublicEnumerableExtensions.Project[TItem,TResult](IEnumerable`1 items, Func`2 projector)+MoveNext()
[24.04.2019 11:13:28 Informational] [xUnit.net 00:00:01.09]            at AgileObjects.AgileMapper.Extensions.PublicEnumerableExtensions.Filter[TItem](IEnumerable`1 items, Func`2 predicate)+MoveNext()
[24.04.2019 11:13:28 Informational] [xUnit.net 00:00:01.09]            at AgileObjects.AgileMapper.Extensions.PublicEnumerableExtensions.Project[TItem,TResult](IEnumerable`1 items, Func`2 projector)+MoveNext()
[24.04.2019 11:13:28 Informational] [xUnit.net 00:00:01.09]            at System.Collections.Generic.LargeArrayBuilder`1.AddRange(IEnumerable`1 items)
[24.04.2019 11:13:28 Informational] [xUnit.net 00:00:01.09]            at System.Collections.Generic.EnumerableHelpers.ToArray[T](IEnumerable`1 source)
[24.04.2019 11:13:28 Informational] [xUnit.net 00:00:01.09]            at System.Linq.Enumerable.ToArray[TSource](IEnumerable`1 source)
[24.04.2019 11:13:28 Informational] [xUnit.net 00:00:01.09]            at AgileObjects.AgileMapper.ObjectPopulation.DictionaryMappingExpressionFactory.GetDictionaryPopulation(IObjectMappingData mappingData)
[24.04.2019 11:13:28 Informational] [xUnit.net 00:00:01.09]            at AgileObjects.AgileMapper.ObjectPopulation.DictionaryMappingExpressionFactory.GetObjectPopulation(MappingCreationContext context)+MoveNext()
[24.04.2019 11:13:28 Informational] [xUnit.net 00:00:01.09]            at AgileObjects.AgileMapper.Extensions.PublicEnumerableExtensions.Filter[TItem](IEnumerable`1 items, Func`2 predicate)+MoveNext()
[24.04.2019 11:13:28 Informational] [xUnit.net 00:00:01.09]            at System.Collections.Generic.List`1.AddEnumerable(IEnumerable`1 enumerable)
[24.04.2019 11:13:28 Informational] [xUnit.net 00:00:01.09]            at System.Collections.Generic.List`1.InsertRange(Int32 index, IEnumerable`1 collection)
[24.04.2019 11:13:28 Informational] [xUnit.net 00:00:01.09]            at AgileObjects.AgileMapper.ObjectPopulation.MappingExpressionFactoryBase.AddPopulationsAndCallbacks(MappingCreationContext context)
[24.04.2019 11:13:28 Informational] [xUnit.net 00:00:01.09]            at AgileObjects.AgileMapper.ObjectPopulation.MappingExpressionFactoryBase.Create(IObjectMappingData mappingData)
[24.04.2019 11:13:28 Informational] [xUnit.net 00:00:01.09]            at AgileObjects.AgileMapper.ObjectPopulation.ObjectMapperFactory.Create[TSource,TTarget](ObjectMappingData`2 mappingData)
[24.04.2019 11:13:28 Informational] [xUnit.net 00:00:01.09]            at AgileObjects.AgileMapper.ObjectPopulation.ObjectMappingData`2.GetOrCreateMapper()
[24.04.2019 11:13:28 Informational] [xUnit.net 00:00:01.09]            at AgileObjects.AgileMapper.Caching.ArrayCache`2.GetOrAdd(TKey key, Func`2 valueFactory)
[24.04.2019 11:13:28 Informational] [xUnit.net 00:00:01.09]            at AgileObjects.AgileMapper.ObjectPopulation.ObjectMapperFactory.GetOrCreateRoot[TSource,TTarget](ObjectMappingData`2 mappingData)
[24.04.2019 11:13:28 Informational] [xUnit.net 00:00:01.09]            at AgileObjects.AgileMapper.ObjectPopulation.ObjectMappingData`2..ctor(TSource source, TTarget target, Nullable`1 enumerableIndex, MappingTypes mappingTypes, IMappingContext mappingContext, IObjectMappingData declaredTypeMappingData, IObjectMappingData parent, Boolean createMapper)
[24.04.2019 11:13:28 Informational] [xUnit.net 00:00:01.09]            at AgileObjects.AgileMapper.ObjectPopulation.ObjectMappingDataFactory.ForRootFixedTypes[TSource,TTarget](TSource source, TTarget target, MappingTypes mappingTypes, IMappingContext mappingContext, Boolean createMapper)
[24.04.2019 11:13:28 Informational] [xUnit.net 00:00:01.09]            at AgileObjects.AgileMapper.ObjectPopulation.ObjectMappingDataFactory.ForRootFixedTypes[TSource,TTarget](TSource source, TTarget target, IMappingContext mappingContext, Boolean createMapper)
[24.04.2019 11:13:28 Informational] [xUnit.net 00:00:01.09]            at AgileObjects.AgileMapper.MappingExecutor`1.PerformMapping[TTarget](TTarget target)
[24.04.2019 11:13:28 Informational] [xUnit.net 00:00:01.09]            at AgileObjects.AgileMapper.MappingExecutor`1.ToANew[TResult]()
```

However, the mapping works correctly, if I remove the `Target.Wrapper` instance and directly map `Dictionary<string, Source.Data>` to `Dictionary<string, Target.Data>`.

**Comment by SteveWilkes** [comment]
I've fixed this in the 1.4 preview branch, release to follow :)

Just to note, mapping `Wrapper` to a Dictionary with this configuration also creates a target `Data` instance using the contents of `Wrapper.Map`. You therefore end up with three entries:

- 'Map', with `Data.Name` set to null (as the source dictionary doesn't contain a 'Name' entry)
- 'first' with `Data.Name` set to 'First'
- 'second' with `Data.Name` set to 'Second'

Thanks for letting me know about the bug! :)

**Comment by bslbckr** [comment]
Thanks for the fix and the hint concerning mapping of `Map` to the resulting `Dictionary`. 
For what reasons is the member `Wrapper.Map` to the resulting `Dictionary`? My understanding of the configuration is, that `Map(ctxt => ctxt.Source.Map).ToTarget()` somehow changes the mapping's source.

Is there a way to avoid this behavior using a different configuration? I've seen that the To- and FromDictionary-Mapping-Configurations allow for giving conditions on how to handle special keys. cf https://agilemapper.readthedocs.io/configuration/Dictionary-Mapping/

**Comment by SteveWilkes** [comment]
`.ToTarget()` was added to address [this request](https://github.com/agileobjects/AgileMapper/issues/64), where with (e.g.) this source:

```C#
class A
{
    public string Value { get; set; }
    public B SubProperty { get; set; }
}
```

...the members belonging to `SubProperty` can be mapped to the target after the members belonging to `A`. I extended its use for specifying the source member to use when the target can't be constructed using the 'parent' type, but that's the only case where it 'switches' the mapping's source.

The easiest way to not have a 'Map' entry is to use:

```C#
Mapper.WhenMapping
    .From<Source.Wrapper>().ToDictionaries()
    .Map(ctxt => ctxt.Source.Map)
    .ToTarget()
    .And.IgnoreTargetMembersWhere(tm => tm.Name == "Map");
```

I've considered having a 'mapping source switching' configuration, like:

```C#
Mapper.WhenMapping
    .From<Source.Wrapper>().ToDictionaries()
    .Map(ctxt => ctxt.Source.Map)
    .ToTargetInstead()
```

...but that'll be non-trivial, and requires a bit more thinking about :)

**Comment by SteveWilkes** [comment]
The fix for this is now available in v1.4-preview3 [on NuGet](https://www.nuget.org/packages/AgileObjects.AgileMapper).

Thanks again! :)
